File: 965.univalued-binary-tree.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution {
    /**
     * @param TreeNode $root
     * @return Boolean
     */
    function isUnivalTree($root) {
        if (empty($root)) {
            return true;
        }
        return $this->isUnivalTreeHelp($root, $root->val);
    }

    function isUnivalTreeHelp($root, $val) {
        if (empty($root)) {
            return true;
        }

        if ($root->val != $val) {
            return false;
        }

        return $this->isUnivalTreeHelp($root->left, $val) && $this->isUnivalTreeHelp($root->right, $val);
    }
}

File: 965.univalued-binary-tree.cpp
// CreateTime: 2019-12-09 00:28:12
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isUnivalTree(TreeNode* root) {
        if (root == NULL) {
            return true;
        }

        return isUnivalTree(root, root->val);
    }

    bool isUnivalTree(TreeNode* root, int val) {
        if (root == NULL) {
            return true;
        }

        if (root->val != val) {
            return false;
        }

        return isUnivalTree(root->left, val) && isUnivalTree(root->right, val);
    }
};
